src/methods.jl

```diff
@@ -175,68 +214,108 @@ FiniteDifferenceMethod:
   coefficients:          [0.08333333333333333, -0.6666666666666666, 0.0, 0.6666666666666666, -0.08333333333333333]
 
 julia> fdm(sin, 1, 1e-3)
-0.5403023058679624
+ 0.5403023058679624
 
 julia> fdm(sin, 1, 1e-3) - cos(1)  # Check the error.
 -1.7741363933510002e-13
 ```
 """
-@inline function (m::FiniteDifferenceMethod)(f::Function, x::Real, h::Real)
-    # Assume that converting to float is desired.
-    return _eval_method(m, f, float(x), h)
+# We loop over all concrete subtypes of `FiniteDifferenceMethod` for 1.0 compatibility.
+for T in (UnadaptedFiniteDifferenceMethod, AdaptedFiniteDifferenceMethod)
+    @eval begin
+        function (m::$T{P,Q})(f::TF, x::Real, step::Real) where {P,Q,TF<:Function}
+            # Assume that converting to float is desired.
+            x = float(x)
+            fs = _eval_function(m, f, x, step)
+            return _compute_estimate(m, fs, x, step, m.coefs)
+        end
+    end
 end
-@inline function _eval_method(
-    m::FiniteDifferenceMethod,
-    f::Function,
+
+function _eval_function(
+    m::FiniteDifferenceMethod, f::TF, x::T, step::Real,
+) where {TF<:Function,T<:AbstractFloat}
+    return f.(x .+ T(step) .* m.grid)
+end
+
+function _compute_estimate(
+    m::FiniteDifferenceMethod{P,Q},
+    fs::SVector{P,TF},
     x::T,
-    h::Real
-) where T<:AbstractFloat
-    return sum(
-        i -> convert(T, m.coefs[i]) * f(T(x + h * m.grid[i])),
-        eachindex(m.grid)
-    ) / h^m.q
+    step::Real,
+    coefs::SVector{P,Float64},
+) where {P,Q,TF,T<:AbstractFloat}
+    # If we substitute `T.(coefs)` in the expression below, then allocations occur. We
+    # therefore perform the broadcasting first.
+    coefs = T.(coefs)
+    return sum(fs .* coefs) ./ T(step)^Q
 end
 
 # Check the method and derivative orders for consistency.
 function _check_p_q(p::Integer, q::Integer)
     q >= 0 || throw(DomainError(q, "order of derivative (`q`) must be non-negative"))
     q < p || throw(DomainError(
         (q, p),
-        "order of the method (q) must be strictly greater than that of the derivative (p)",
+        "order of the method (`p`) must be strictly greater than that of the derivative " *
+        "(`q`)",
     ))
-    # Check whether the method can be computed. We require the factorial of the method order
-    # to be computable with regular `Int`s, but `factorial` will after 20, so 20 is the
-    # largest we can allow.
-    p > 20 && throw(DomainError(p, "order of the method (`p`) is too large to be computed"))
-    return
 end
 
-const _COEFFS_CACHE = Dict{Tuple{AbstractVector{<:Real}, Integer}, Vector{Float64}}()
+function _coefs(grid, p, q)
+    # For high precision on the `\`, we use `Rational`, and to prevent overflows we use
+    # `BigInt`. At the end we go to `Float64` for fast floating point math, rather than
+    # rational math.
+    C = [Rational{BigInt}(g^i) for i in 0:(p - 1), g in grid]
```

Comment:
was `Int128` not doing it for us?

Reply:
No, `Int128` unfortunately hit overflow for orders bigger than 12.